nibabel/brikhead.py

```diff
@@ -0,0 +1,488 @@
+            (``None``) will result in the value of
+            ``nibabel.arrayproxy.KEEP_FILE_OPEN_DEFAULT` being used.
+        """
+        super(AFNIArrayProxy, self).__init__(file_like,
+                                             header,
+                                             mmap=mmap,
+                                             keep_file_open=keep_file_open)
+        self._scaling = header.get_data_scaling()
+
+    @property
+    def scaling(self):
+        return self._scaling
+
+    def __array__(self):
+        raw_data = self.get_unscaled()
+        # apply volume specific scaling (may change datatype!)
+        if self._scaling is not None:
+            return raw_data * self._scaling
+
+        return raw_data
+
+    def __getitem__(self, slicer):
+        raw_data = super(AFNIArrayProxy, self).__getitem__(slicer)
+        # apply volume specific scaling (may change datatype!)
+        if self._scaling is not None:
+            scaling = self._scaling.copy()
+            fake_data = strided_scalar(self._shape)
+            _, scaling = np.broadcast_arrays(fake_data, scaling)
+            raw_data = raw_data * scaling[slicer]
+
+        return raw_data
+
+
+class AFNIHeader(SpatialHeader):
+    """Class for AFNI header"""
+    def __init__(self, info):
+        """
+        Parameters
+        ----------
+        info : dict
+            Information from HEAD file as obtained by `parse_AFNI_header()`
+
+        Examples
+        --------
+        >>> header = AFNIHeader(parse_AFNI_header('example4d+orig.HEAD'))
+        >>> header.get_data_dtype()
+        dtype('int16')
+        >>> header.get_zooms()
+        (3.0, 3.0, 3.0, 3.0)
+        >>> header.get_data_shape()
+        (33, 41, 25, 3)
+        """
+        self.info = info
+        dt = _get_datatype(self.info)
+        super(AFNIHeader, self).__init__(data_dtype=dt,
+                                         shape=self._calc_data_shape(),
+                                         zooms=self._calc_zooms())
+
+    @classmethod
+    def from_header(klass, header=None):
+        if header is None:
+            raise AFNIError('Cannot create AFNIHeader from nothing.')
+        if type(header) == klass:
+            return header.copy()
+        raise AFNIError('Cannot create AFNIHeader from non-AFNIHeader.')
+
+    @classmethod
+    def from_fileobj(klass, fileobj):
+        info = parse_AFNI_header(fileobj)
+        return klass(info)
+
+    def copy(self):
+        return AFNIHeader(deepcopy(self.info))
+
+    def _calc_data_shape(self):
+        """Calculate the output shape of the image data
+
+        Returns length 3 tuple for 3D image, length 4 tuple for 4D.
+
+        Returns
+        -------
+        (x, y, z, t) : tuple of int
+        """
+        dset_rank = self.info['DATASET_RANK']
+        shape = tuple(self.info['DATASET_DIMENSIONS'][:dset_rank[0]])
+        n_vols = dset_rank[1]
+
+        return shape + (n_vols,) if n_vols > 1 else shape
```

Comment:
Well, no. By default, AFNI datasets are actually always ndim=4 of shape (x, y, z, >=1), because of how the dimensions are stored in the HEAD file (see response below). You'll never get an (x, y, z) dataset because that would preclude you from subsetting along the "time" axis, and AFNI programs feature heavily in permitting you to do this. I'll make sure to remedy this in my next update.